src/permanent.cpp

```diff
@@ -0,0 +1,269 @@
+ * @param cols A vector representing the multiplicity of each column.
+ * @return The permanent of the input matrix.
+ */
+template <typename T>
+std::complex<T> permanent_cpp(Matrix<std::complex<T>> &A, Vector<int> &rows, Vector<int> &cols)
+{
+    using TComplex = std::complex<T>;
+
+    size_t min_idx = 0;
+    int minelem = 0;
+
+    for (size_t i = 0; i < rows.size(); i++)
+    {
+        if (minelem == 0 || (rows[i] < minelem && rows[i] != 0))
+        {
+            minelem = rows[i];
+            min_idx = i;
+        }
+    }
+
+    if (rows.size() > 0 && minelem != 0)
+    {
+        Vector<int> rows_(rows.size() + 1);
+
+        rows_[0] = 1;
+
+        for (size_t i = 0; i < rows.size(); i++)
+        {
+            rows_[i + 1] = rows[i];
+        }
+        rows_[1 + min_idx] -= 1;
+
+        Matrix<TComplex> mtx_(A.rows + 1, A.cols);
+
+        for (size_t j = 0; j < A.cols; j++)
+        {
+            mtx_(0, j) = A(min_idx, j);
+        }
+
+        for (size_t i = 0; i < A.rows; i++)
+        {
+            for (size_t j = 0; j < A.cols; j++)
+            {
+                mtx_(i + 1, j) = A(i, j);
+            }
+        }
+
+        rows = rows_;
+        A = mtx_;
+    }
+
+    int sum_rows = rows.sum();
+    int sum_cols = cols.sum();
+
+    if (sum_rows != sum_cols)
+    {
+        std::string error("Number of input and output states should be equal");
+        throw error;
+    }
+
+    if (A.rows == 0 || A.cols == 0 || sum_rows == 0 || sum_cols == 0)
+        // the permanent of an empty matrix is 1 by definition
+        return TComplex(1.0, 0.0);
+
+    if (A.rows == 1)
+    {
+        TComplex ret(1.0, 0.0);
+        for (size_t i = 0; i < cols.size(); i++)
+        {
+            for (int j = 0; j < cols[i]; j++)
+            {
+                ret *= A[i];
+            }
+        }
+
+        return TComplex(ret.real(), ret.imag());
+    }
+
+    Matrix<TComplex> mtx2(A.rows, A.cols);
+    for (size_t i = 0; i < A.size(); i++)
+    {
+        mtx2[i] = A[i] * TComplex(2.0, 0.0);
+    }
+
+    // Create a dynamically allocated array for n-ary limits.
+    size_t n_ary_size = rows.size() - 1;
+    int *n_ary_limits = new int[n_ary_size];
+    for (size_t i = 0; i < n_ary_size; i++)
+    {
+        n_ary_limits[i] = rows[i + 1] + 1;
+    }
+
+    int64_t idx_max = n_ary_limits[0];
+    for (size_t i = 1; i < n_ary_size; i++)
+    {
+        idx_max *= n_ary_limits[i];
+    }
+
+    // determine the concurrency of the calculation
+    unsigned int n_threads = std::thread::hardware_concurrency();
+    int64_t concurrency = static_cast<int64_t>(n_threads) * 4;
+    concurrency = concurrency < idx_max ? concurrency : static_cast<int64_t>(idx_max);
+
+    std::vector<TComplex> thread_results(static_cast<unsigned int>(concurrency), TComplex(0.0, 0.0));
+
+// See: https://stackoverflow.com/a/8447025
+#if defined(_OPENMP)
+#pragma omp parallel for num_threads(concurrency)
```

Comment:
There is a new error here on Windows, similar to the previous comment. Is there a reason to use `int64_t` for `work_batch`, can't we just use `uint64_t` instead, and for all the variables derived from this?